### PR TITLE
Fix missing subclass discriminator in ON clause for entity joins

### PR DESCRIPTION
**Pull request by bahusoid**
This issue affects both Criteria/QueryOver and Hql/Linq Query API. While it's regression only for LINQ case I made fix also for Criteria (as fix is quite simple).

Fixes https://github.com/nhibernate/nhibernate-core/issues/2599  (complete fix for #2463)

### CODE REVIEW

File: src/NHibernate.Test/NHSpecificTest/GH2463/Fixture.cs [hunk 1]
```diff
@@ -12,18 +13,58 @@ protected override string[] Mappings
 			get { return new[] {"ABC.hbm.xml"}; }
 		}
 
+		protected override void OnSetUp()
+		{
+			using (var session = OpenSession())
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/NHSpecificTest/GH2463/Fixture.cs:18]: This test fixture should be ignored if `!Dialect.SupportsScalarSubSelects`

**Review by fredericDelaporte — APPROVED** [review]

**Comment by fredericDelaporte** [comment]
Should we wait a bit in case of other issues waiting to be reported, or release this soon, like on next Sunday?

**Comment by hazzik** [comment]
I would say don't wait.

**Comment by bahusoid** [comment]
Just in case of troubles with resolving conflicts when merge to master:
https://github.com/bahusoid/nhibernate-core/tree/GH2599UP

File: src/NHibernate/Loader/JoinWalker.cs [hunk 2]
```diff
@@ -851,8 +852,8 @@ protected JoinFragment MergeOuterJoins(IList<OuterJoinableAssociation> associati
 				{
 					oj.AddJoins(outerjoin);
 					// NH Different behavior : NH1179 and NH1293
-					// Apply filters in Many-To-One association
-					if (enabledFiltersForManyToOne.Count > 0)
+					// Apply filters for entity joins and Many-To-One associations
+					if (oj.ForceFilter || enabledFiltersForManyToOne.Count > 0)
 					{
 						string manyToOneFilterFragment = oj.Joinable.FilterFragment(oj.RHSAlias, enabledFiltersForManyToOne);
```
Comment (fredericDelaporte) [inline src/NHibernate/Loader/JoinWalker.cs:858]: Similarly:
```
					// Apply filters for entity joins and Many-To-One associations
					var enabledFiltersForJoin = oj.ForceFilter ? enabledFilters : enabledFiltersForManyToOne;
					if (oj.ForceFilter || enabledFiltersForJoin.Count > 0)
					{
						string manyToOneFilterFragment = oj.Joinable.FilterFragment(oj.RHSAlias, enabledFiltersForJoin);
```

**Pull request merged**